### ISSUE TITLE
Correct relevance values?

### ISSUE THREAD

**Issue by wmelton** [tensorflow/ranking#90]
Hello, 

If we have millions of user events that are mapped to relevance judgments for a given <query, document> pair, how does TF-Rank expect the LibSVM formatted file to be presented?

Meaning - is best practice to average relevance across all user events such that each <query, document> pair only has one row in the LibSVM file, or would a <query, document> pair exist in LibSVM file for every single user who interacted with that document in some way (millions of rows for a given <query, document> pair)? 

If a row should exist for every user interaction, would we encode user id as a normalized value in the embedding at that point?

Not clear from the documentation that exists what best practice is - appreciate any insight.

Thank you in advance.

**Comment by wmelton** [comment]
@dynamicwebpaige - do you happen to know the answer to this one?

**Comment by ramakumar1729** [comment]
@wmelton The number of documents per row/per query, are the ones compared by the user to produce relevance judgments. You don't need to average them, you can treat each of them as separate queries. 

For user ID, a sparse feature, you can either use an embedding based representation, or handle it as a sparse feature in the computational graph via tf.feature_column. Look at [git.io/tf-ranking-demo](git.io/tf-ranking-demo) for how to incorporate sparse features into TF-Ranking model.

**Comment by wmelton** [comment]
@ramakumar1729 - thank you so much for your response. 

For clarity, I'm looking at some of the training data in the example exercises, and I don't see any documents that have multiple relevance scores for the same document - 

https://raw.githubusercontent.com/tensorflow/ranking/master/tensorflow_ranking/examples/data/train.txt

Are you saying the following would be acceptable? 

```
0 qid:1 115:-0.617277 5:-0.109573 n:... #DocId 123
2 qid:1 75:-0.332035 84:0.926297 n:... #DocId 456
0 qid:1 75:-0.332035 84:0.926297 n:... #DocId 456
5 qid:1 75:-0.332035 84:0.926297 n:... #DocId 456
1 qid:1 75:-0.332035 84:0.926297 n:... #DocId 456
2 qid:1 100:0.871419 118:-0.494900 n... #DocId 246
```

Where DocId 456 in the example above would have different relevance judgments based on different users assessments of relevance? 

Would this work even if no user data was supplied and the embeddings for each document were identical?

**Comment by ramakumar1729** [comment]
In terms of format, this is fine. From a modelling perspective, having two datapoints with identical features and different labels is not useful. You will need to add features that helps the model discriminate between the positive labels and the negative labels.

**Comment by wmelton** [comment]
@ramakumar1729 I think I'm tracking with you now. I'm assuming any features that represent categorical information would need to be normalized in this case, even in a sparse vector format? Also assuming its acceptable for certain sparse vector values to be absent on the inference side that were present during model training?

**Comment by ramakumar1729** [comment]
Hi Wes, sorry for the delay, I was traveling. I'm not sure what you mean by normalizing sparse features. Please check out this tutorial on handling sparse features using TF-Ranking: [git.io/tf-ranking-demo](git.io/tf-ranking-demo).

A missing sparse feature would be treated as an empty string (you can change this via feature column specification). Hence it will be acceptable for the feature to be missing as long as you have a default value.

**Comment by ramakumar1729** [comment]
Closing this for now.
